### PR TITLE
planner: when column collation is consistent, constant propagation is required | tidb-test=pr/2271

### LINKED ISSUE #47629: collation_connection affects the passing of constants cause wrong query plan
## Bug Report

collation_connection affects the passing of constants cause wrong query plan
### 1. Minimal reproduce step (Required)
```
CREATE TABLE `t2` (
  `id1` char(8) NOT NULL COMMENT '编号',
  `id2` varchar(180) NOT NULL COMMENT '姓名',
  `id3` varchar(12) NOT NULL COMMENT '岗位编号',
  `id4` char(9) NOT NULL COMMENT '机构编号',
  `id5` varchar(10) NOT NULL COMMENT '角色编号',
  `id6` varchar(60) DEFAULT NULL COMMENT '登录名',
  PRIMARY KEY (`id1`,`id2`,`id3`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
CREATE TABLE `t1` (
  `id1` varchar(200) NOT NULL COMMENT '编号',
  `id2` varchar(200) DEFAULT NULL COMMENT '账号',
  `id3` varchar(600) DEFAULT NULL COMMENT '姓名',
  `id4` varchar(600) DEFAULT NULL COMMENT '电话',
  PRIMARY KEY (`id1`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
set collation_connection=utf8_general_ci;

```

### 2. What did you expect to see? (Required)
```
mysql> explain format='verbose'  SELECT  t.id1, t.id2, t.id3 FROM t1 t
    ->  JOIN t2
    -> ON t.id1 = t2.id1
    -> WHERE t2.id1 = '63812828';
+-----------------------------+---------+---------+-----------+----------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| id                          | estRows | estCost | task      | access object                          | operator info                                                                                                           |
+-----------------------------+---------+---------+-----------+----------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| IndexJoin_12                | 1.25    | 38.15   | root      |                                        | inner join, inner:IndexReader_11, outer key:test.t1.id1, inner key:test.t2.id1, equal cond:eq(test.t1.id1, test.t2.id1) |
| ├─Point_Get_38(Build)   | 1.00    | 5.60    | root      | table:t1, index:PRIMARY(id1)           |                                                                                                                         |
| └─IndexReader_11(Probe) | 1.25    | 9.11    | root      |                                        | index:IndexRangeScan_10                                                                                                 |
|   └─IndexRangeScan_10   | 1.25    | 0.00    | cop[tikv] | table:t2, index:PRIMARY(id1, id2, id3) | range: decided by [eq(test.t2.id1, test.t1.id1)], keep order:false, stats:pseudo                                        |
+-----------------------------+---------+---------+-----------+----------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
4 rows in set, 1 warning (0.00 sec)
```
### 3. What did you see instead (Required)
```
mysql> explain format='verbose'  SELECT /*+ INL_JOIN(t2) */ t.id1, t.id2, t.id3 FROM t1 t use index(PRIMARY)
    ->  JOIN t2
    -> ON t.id1 = t2.id1
    -> WHERE t2.id1 = '63812828';
+------------------------------------+----------+-----------+-----------+----------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| id                                 | estRows  | estCost   | task      | access object                          | operator info                                                                                                           |
+------------------------------------+----------+-----------+-----------+----------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| IndexJoin_12                       | 12.50    | 399770.41 | root      |                                        | inner join, inner:IndexReader_11, outer key:test.t1.id1, inner key:test.t2.id1, equal cond:eq(test.t1.id1, test.t2.id1) |
| ├─IndexLookUp_22(Build)        | 10000.00 | 229828.94 | root      |                                        |                                                                                                                         |
| │ ├─IndexFullScan_20(Build)  | 10000.00 | 555000.00 | cop[tikv] | table:t, index:PRIMARY(id1)            | keep order:false, stats:pseudo                                                                                          |
| │ └─TableRowIDScan_21(Probe) | 10000.00 | 555000.00 | cop[tikv] | table:t                                | keep order:false, stats:pseudo                                                                                          |
| └─IndexReader_11(Probe)        | 0.00     | 7.34      | root      |                                        | index:Selection_10                                                                                                      |
|   └─Selection_10               | 0.00     | 0.00      | cop[tikv] |                                        | eq(test.t2.id1, "63812828")                                                                                             |
|     └─IndexRangeScan_9         | 1.25     | 0.00      | cop[tikv] | table:t2, index:PRIMARY(id1, id2, id3) | range: decided by [eq(test.t2.id1, test.t1.id1)], keep order:false, stats:pseudo                                        |
+------------------------------------+----------+-----------+-----------+----------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
7 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

v6.1.1

### PR DESCRIPTION
**Pull request by King-Dylan**
### What problem does this PR solve?
Issue Number: close #47629   
Problem Summary:
When the collation set by collation_connection is inconsistent with the column collation, it may affect the normal propagation of constants.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
When the collation_connection setting conflicts with column collation, this release ensures consistent handling for proper constant propagation and introduces compatibility improvements, aligning with MySQL 8.0 standards.
```

### CODE REVIEW

**Comment by King-Dylan** [comment]
fix mysql-test: https://github.com/PingCAP-QE/tidb-test/pull/2271

**Comment by King-Dylan** [comment]
After this pull request, it will change the behavior for the following query, but it will be compatible with MySQL 8.0
`mysql> create table t (a char(10) collate utf8mb4_general_ci, b char(10) collate utf8mb4_general_ci);
Query OK, 0 rows affected (0.10 sec)
mysql> insert into t values ('A', 'a');
Query OK, 1 row affected (0.01 sec)
mysql> select * from t t1, t t2 where t1.a=t2.b and t2.b='a' collate utf8mb4_bin;
Empty set (0.00 sec)
mysql> select * from t t1, t t2 where t1.a=t2.b and t2.b='a';
+------+------+------+------+
| a    | b    | a    | b    |
+------+------+------+------+
| A    | a    | A    | a    |
+------+------+------+------+
1 row in set (0.00 sec)`

**Comment by bb7133** [comment]
Please update the 'Release Note' for the changes.

**Comment by wjhuang2016** [comment]
> After this pull request, it will change the behavior for the following query, but it will be compatible with MySQL 8.0
> 
> ```
> mysql> create table t (a char(10) collate utf8mb4_general_ci, b char(10) collate utf8mb4_general_ci);
> Query OK, 0 rows affected (0.10 sec)
> mysql> insert into t values ('A', 'a');
> Query OK, 1 row affected (0.01 sec)
> mysql> select * from t t1, t t2 where t1.a=t2.b and t2.b='a' collate utf8mb4_bin;
> Empty set (0.00 sec)
> mysql> select * from t t1, t t2 where t1.a=t2.b and t2.b='a';
> +------+------+------+------+
> | a    | b    | a    | b    |
> +------+------+------+------+
> | A    | a    | A    | a    |
> +------+------+------+------+
> 1 row in set (0.00 sec)
> ```

In MySQL:
``` SQL
mysql> select * from t t1, t t2 where t1.a=t2.b and t2.b='a' collate utf8mb4_bin;
+------+------+------+------+
| a    | b    | a    | b    |
+------+------+------+------+
| A    | a    | A    | a    |
+------+------+------+------+
1 row in set (0.01 sec)

mysql> select * from t t1, t t2 where t1.a=t2.b and t2.b='a';
+------+------+------+------+
| a    | b    | a    | b    |
+------+------+------+------+
| A    | a    | A    | a    |
+------+------+------+------+
1 row in set (0.00 sec)
```

File: pkg/expression/util.go [hunk 1]
```diff
@@ -494,7 +494,10 @@ func ColumnSubstituteImpl(ctx sessionctx.Context, expr Expression, schema *Schem
 					logutil.BgLogger().Error("Unexpected error happened during ColumnSubstitution", zap.Stack("stack"))
 					return false, failed, v
 				}
-				if oldCollEt.Collation == newCollEt.Collation {
+				// When column collations are consistent, constant propagation can proceed without checking the collation of the constants.
+				if len(v.GetArgs()) == 2 && v.GetArgs()[0].GetType().GetCollate() == v.GetArgs()[1].GetType().GetCollate() {
+					changed = true
+				} else if oldCollEt.Collation == newCollEt.Collation {
```
Comment (AilinKid) [inline pkg/expression/util.go:500]: image
from [link](https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html) here, collation_connection does affect constant value's collation, while mysql plan indeed gives a quick return for one side of join is empty table (similar as quick point get and found nothing)

```
mysql> explain  SELECT  t.id1, t.id2, t.id3 FROM t1 t JOIN t2 ON t.id1 = t2.id1 WHERE t2.id1 = '63812828';
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                          |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------+
|  1 | SIMPLE      | NULL  | NULL       | NULL | NULL          | NULL | NULL    | NULL | NULL |     NULL | no matching row in const table |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+--------------------------------+
1 row in set, 1 warning (0.03 sec)
```
my advices is that you can dive a liiter bit of mysql opt code, figure out how them handle column's collation itself and those collation from constant value inherited from collation_connection

**Pull request closed without merging**